File: app/src/main/java/com/lence/penguinsvskillerwhales/life/Move.java
package com.lence.penguinsvskillerwhales.life;

import android.graphics.Point;

import com.lence.penguinsvskillerwhales.utils.Constants;
import com.lence.penguinsvskillerwhales.view.KotlinField;
import com.lence.penguinsvskillerwhales.view.UpdatePresenter;

import java.util.Random;

public class Move {
    private UpdatePresenter mUpdatePresenter;
    private final KotlinField mField;
    private final Point mCurrent;

    Move(UpdatePresenter updatePresenter, KotlinField field, Point mCurrent) {
        mUpdatePresenter = updatePresenter;
        mField = field;
        this.mCurrent = mCurrent;
    }


    public void execute() {
        Random random = new Random();
        int randomInt = random.nextInt(Constants.count_choice);

        if (findDirection(randomInt)) {
            move(KotlinField.getPointByDirection(mCurrent.x,mCurrent.y,randomInt));
        }

        mUpdatePresenter.update(mField);
    }


    private void move(Point to) {
        mField.set(to.x, to.y, mField.get(mCurrent.x, mCurrent.y));
        mField.set(mCurrent.x, mCurrent.y, null);
        //Log.e("move", "move " + to + " to right down");
    }

    private boolean findDirection(int direction) {
        Point p = KotlinField.getPointByDirection(mCurrent.x,mCurrent.y, direction);
        //Log.e("point",p +" "+ mField.isInField(p.x, p.y));
        return (mField.isInField(p.x, p.y) && (mField.get(p.x, p.y) == null));

    }



}

File: app/src/main/java/com/lence/penguinsvskillerwhales/life/Move.kt
package com.lence.penguinsvskillerwhales.life

import android.graphics.Point
import com.lence.penguinsvskillerwhales.utils.Constants
import com.lence.penguinsvskillerwhales.view.Field
import com.lence.penguinsvskillerwhales.view.UpdatePresenter
import java.util.*

class Move internal constructor(private val mUpdatePresenter: UpdatePresenter,
                                private val mField: Field,
                                private val mCurrent: Point) {


    fun execute() {
        val random = Random()
        val randomInt = random.nextInt(Constants.count_choice)

        if (findDirection(randomInt)) {
            move(Field.getPointByDirection(mCurrent.x, mCurrent.y, randomInt))
        }

        mUpdatePresenter.update(mField)
    }


    private fun move(to: Point) {
        mField[to.x, to.y] = mField[mCurrent.x, mCurrent.y]
        mField[mCurrent.x, mCurrent.y] = null
        //Log.e("move", "move " + to + " to right down");
    }

    private fun findDirection(direction: Int): Boolean {
        val p = Field.getPointByDirection(mCurrent.x, mCurrent.y, direction)
        //Log.e("point",p +" "+ mField.isInField(p.x, p.y));
        return mField.isInField(p.x, p.y) && mField[p.x, p.y] == null

    }


}
